Fixes #38173 - Permit use of arrays for multiple timesources

jcpunk opened the pull request:
I'm unclear how best to test this.

My site offers multiple on site NTP servers on different names depending on location.  This permits them to bring hosts down for maintenance without significant interruption.

ShimShtein commented on app/views/unattended/provisioning_templates/provision/kickstart_default.erb:
```diff
@@ -142,13 +142,27 @@ authselect --useshadow --passalgo=<%= @host.operatingsystem.password_hash.downca
 <% if rhel_compatible && os_major < 9 -%>
-timezone --utc <%= host_param('time-zone') || 'UTC' %> <%= host_param('ntp-server') ? "--ntpservers #{host_param('ntp-server')}" : '' %>
+<% if host_param('ntp-server') -%>
+<% if host_param('ntp-server').is_a?(Array) -%>
```
Since `is_a?(Array)` is not allowed in safemode, you can replace it with `is_a?(String)` which is used in other templates too.

jcpunk replied:
In theory this is sorted.

ShimShtein reviewed (commented):
>I'm unclear how best to test this.

You can add the parameter to https://github.com/theforeman/foreman/blob/62b3353e6b557c70f4e1e51f6d2b0de48001c119/app/services/foreman/template_snapshot_service.rb#L68

ekohl reviewed (commented):
I'm not sure if safemode allows it, but there's a pattern that may help: `Array('x') == Array(['x'])`. So in practice that means you should be able to do:
```ruby
Array(host_param('ntp-server')).join(',')
```

And:
```ruby
Array(host_param('ntp-server')).each do |ntpserver|
```

Note that `Array(nil) == []` so you can avoid the condition.

However, it looks like safemode doesn't allow `Array` to be used. Perhaps we should.

jcpunk commented:
> You can add the parameter to [foreman/app/services/foreman/template_snapshot_service.rb](https://github.com/theforeman/foreman/blob/62b3353e6b557c70f4e1e51f6d2b0de48001c119/app/services/foreman/template_snapshot_service.rb#L68)

I'm not seeing where this has an existing test for when it is of type `String`.  I'd expect for coverage and sanity it should check both `String` and `Array` types to make sure each kind works as expected.

Testwise, I'm not sure switching to `is_a(String)` seems to be on the safemode list?

ekohl commented:
> I'm unclear how best to test this.

In https://github.com/theforeman/foreman/pull/10605#pullrequestreview-3009592626 I described where we provide parameters. Then it runs through the snapshots and is validated.

In that PR I already felt the need for a "plain" (all defaults) host and then introduce another host with non-default values. Perhaps that's something you could look at, given how this code is non-trivial.

For this specific case I'd be tempted to not store the full snapshot but have a dedicated test with asserts. https://github.com/theforeman/foreman/commit/d7e1c39909cf365dcddf3b8ffff7b2c15c88377a can be another inspiration for that. You can see that in `test/unit/foreman/templates` we already have some template testing infrastructure.

jcpunk commented:
I've tried to add in test conditions.  Per my usual behavior, `RAILS_ENV=test bundle exec rails snapshots:generate` provides a series of errors that make no sense to me.

```shell
RAILS_ENV=test bundle exec rails snapshots:generate
rails aborted!
ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  relation "settings" does not exist
LINE 1: SELECT "settings".* FROM "settings"
                                 ^
/data/app/registries/setting_registry.rb:183:in `load_values'
/data/app/registries/setting_registry.rb:161:in `load'
/data/lib/tasks/snapshots.rake:34:in `block (3 levels) in <main>'
/data/lib/tasks/snapshots.rake:33:in `block (2 levels) in <main>'
/data/bin/rails:9:in `<top (required)>'
/data/bin/spring:12:in `require'
/data/bin/spring:12:in `<top (required)>'

Caused by:
PG::UndefinedTable: ERROR:  relation "settings" does not exist
LINE 1: SELECT "settings".* FROM "settings"
                                 ^
/data/app/registries/setting_registry.rb:183:in `load_values'
/data/app/registries/setting_registry.rb:161:in `load'
/data/lib/tasks/snapshots.rake:34:in `block (3 levels) in <main>'
/data/lib/tasks/snapshots.rake:33:in `block (2 levels) in <main>'
/data/bin/rails:9:in `<top (required)>'
/data/bin/spring:12:in `require'
/data/bin/spring:12:in `<top (required)>'
Tasks: TOP => snapshots:generate
(See full trace by running task with --trace)
```

`db:create` and `db:migrate` are similarly erroring out.

ShimShtein reviewed (commented):
Can you please rebase this PR to make sure the tests run?

Besides that I see there are more usages of the ntp-server paramter across the code:
https://github.com/theforeman/foreman/blob/7dd021f93c661274df4559769c1ea5aebd7fa78e/app/views/unattended/provisioning_templates/provision/xenserver_default_answerfile.erb#L34
https://github.com/theforeman/foreman/blob/7dd021f93c661274df4559769c1ea5aebd7fa78e/app/views/unattended/provisioning_templates/provision/preseed_default.erb#L100
https://github.com/theforeman/foreman/blob/7dd021f93c661274df4559769c1ea5aebd7fa78e/app/views/unattended/provisioning_templates/finish/junos_default_finish.erb#L26
https://github.com/theforeman/foreman/blob/7dd021f93c661274df4559769c1ea5aebd7fa78e/app/views/unattended/provisioning_templates/finish/freebsd_(mfsbsd)_finish.erb#L36

The search that I was using: https://github.com/search?q=repo%3Atheforeman%2Fforeman+ntp+server&type=code

I think we need to adapt all of those places to properly handle the ntp servers case.

jcpunk commented:
Patch rebased.

jcpunk commented:
The only one of these I actually use is anaconda.

I've attempted the other installer tweaks based on their online documentation.

jcpunk commented:
The template generation is producing an error on the `ntp-server` parameter test I added to RHEL.  I have no idea how to fix that...

ShimShtein commented on app/services/foreman/template_snapshot_service.rb:
```diff
@@ -196,6 +200,17 @@ def ubuntu_autoinst4dhcp
+    def rhel9_dhcp_ntp_multiple
+      host = FactoryBot.build(:host_for_snapshots_ipv4_dhcp_ntp_multiple_rhel9,
+        name: 'snapshot-ipv4-dhcp-ntp-multiple-rhel9',
+        'ntp-server': ['127.0.0.1', '192.168.0.0'],
```
`ntp-server` is a host parameter, not a property of the host object. You need to add it with the rest of the parameters in `define_host_params` method: https://github.com/theforeman/foreman/blob/b1343ecd50296d0dddcdee076a4e078236c43f2f/app/services/foreman/template_snapshot_service.rb#L97

jcpunk commented:
I've tried to fix the testing again...

ShimShtein reviewed (dismissed):
LGTM, @ekohl if you're OK with this, I think we can merge.

ekohl commented on app/views/unattended/provisioning_templates/finish/freebsd_(mfsbsd)_finish.erb:
```diff
@@ -33,7 +33,15 @@ echo 'PermitRootLogin yes' >> /etc/ssh/sshd_config
 cp /usr/share/zoneinfo/<%= host_param('time-zone') || 'UTC' %> /etc/localtime
 adjkerntz -a
-ntpdate <%= host_param('ntp-server') || '0.freebsd.pool.ntp.org' %>
+<% if host_param('ntp-server') -%>
+<% if host_param('ntp-server').is_a?(String) -%>
+ntpdate <%= host_param('ntp-server') %>
+<% else -%>
+ntpdate <%= host_param('ntp-server').join(' ') %>
+<% end -%>
+<% else -%>
+ntpdate 0.freebsd.pool.ntp.org
+<% end -%>
```
I'd be a bit more defensive here and swap the condition. The main reason is that `.join` only works on arrays while the other option calls `.to_s`, which is defined on all objects. If the user for some reason inputs something like an integer (or an `IPAddr` instance) the whole template won't crash. While unlikely, I think it's still a good practice.

And I'd also fold the conditions:

```suggestion
<% if host_param('ntp-server').is_a?(Array) -%>
ntpdate <%= host_param('ntp-server').join(' ') %>
<% elsif host_param('ntp-server') -%>
ntpdate <%= host_param('ntp-server') %>
<% else -%>
ntpdate 0.freebsd.pool.ntp.org
<% end -%>
```

When you have that, you can even stick closer to the original code, making the patch even smaller.

```suggestion
<% if host_param('ntp-server').is_a?(Array) -%>
ntpdate <%= host_param('ntp-server').join(' ') %>
<% else -%>
ntpdate <%= host_param('ntp-server') %> || '0.freebsd.pool.ntp.org'
<% end -%>
```

ekohl commented on app/views/unattended/provisioning_templates/provision/preseed_default.erb:
```diff
@@ -97,7 +97,16 @@ d-i time/zone string <%= host_param('time-zone') || 'UTC' %>
+d-i clock-setup/ntp-server string <%= host_param('ntp-server') %>
+<% else -%>
+# host_param('ntp-server') is an array, but preseed only permits one value
+d-i clock-setup/ntp-server string <%= host_param('ntp-server').first %>
```
This surprised me, but you're right that it only takes a single value: https://salsa.debian.org/installer-team/clock-setup/-/blob/master/debian/clock-setup.postinst?ref_type=heads#L25

Now reading this I also wonder if providing a default is good here: we ignore the DHCP provided option. However, that's not in scope for this PR so :+1: for keeping it hardcoded.

ekohl reviewed (changes requested):
I made 1 inline comment, but IMHO it's a repeated pattern. If you swap it around you get the same thing with less code.

jcpunk commented:
Merged suggested change, squashed, and rebased off `develop`.

I'll rework the other `if` blocks.

I'm unsure if I've got something overall rational for `junos` since there is the `server` and `boot-server` bits and I've no experience with that product.

jcpunk commented:
hmmm

```
Safemode doesn't allow to access 'constant' on Array
```

ekohl commented:
Oh that's annoying. Perhaps you can fake it and use `host_param().respond_to?(:join)` to determine if it's an array.

jcpunk commented:
Took a stab at fixing with the check for `join`.

I am staring at the error and not seeing I've done wrongly in the CI...

ShimShtein commented:
@jcpunk can you please rebase this PR? I want to test and merge it.
The code looks quite good, so I don't expect too much rejects on this one.

jcpunk committed:
Fixes #38173 - Permit use of arrays for multiple timesources

Signed-off-by: Pat Riehecky <riehecky@fnal.gov>

jcpunk commented:
Rebased off head

jcpunk commented:
I'm not sure I understand the failure. I'm showing `timesource` is a valid parameter. https://pykickstart.readthedocs.io/en/latest/kickstart-docs.html#timesource

evgeni commented on app/views/unattended/provisioning_templates/provision/kickstart_default.erb:
```diff
@@ -151,17 +151,27 @@ authconfig --useshadow --passalgo=<%= @host.operatingsystem.password_hash.downca
 <% end -%>
+<% elsif host_param('ntp-server').respond_to?(:join) -%>
+<% host_param('ntp-server').each do |ntpserver| -%>
+timesource --ntp-server <%= ntpserver %>
```
timesource is only available in EL9+, but according to the snapshots this is now rendered also for rhel7 and 8

evgeni replied:
oh yeah, here is an indented version which shows it:
```erb
<%# Timezone + NTP -%>
<% if rhel_compatible && os_major < 9 -%>
  <% if host_param('ntp-server').respond_to?(:join) -%>
    timezone --utc <%= host_param('time-zone') || 'UTC' %> --ntpservers <%= host_param('ntp-server').join(',') %>
  <% elsif host_param('ntp-server') -%>
    timezone --utc <%= host_param('time-zone') || 'UTC' %> --ntpservers <%= host_param('ntp-server') %>
  <% else -%>
    timezone --utc <%= host_param('time-zone') || 'UTC' %>
  <% end -%>
<% else -%>
  timezone --utc <%= host_param('time-zone') || 'UTC' %>
<% end -%>

<% if host_param('ntp-pools') -%>
  <% host_param('ntp-pools').each do |ntppool| -%>
    timesource --ntp-pool <%= ntppool %>
  <% end -%>
<% elsif host_param('ntp-server').respond_to?(:join) -%>
  <% host_param('ntp-server').each do |ntpserver| -%>
    timesource --ntp-server <%= ntpserver %>
  <% end -%>
<% elsif host_param('ntp-server') -%>
  timesource --ntp-server <%= host_param('ntp-server') %>
<% end -%>
```

the whole `timesource` part is not in the `else` block of `if rhel_compatible && os_major < 9` anymore

evgeni commented:
we probably should add EL10 snapshots (and consider dropping EL7)

evgeni commented:
> we probably should add EL10 snapshots (and consider dropping EL7)

https://github.com/theforeman/foreman/pull/10819

jcpunk commented:
I cant get the snapshots to generate. It complains the db doesnt exist when I try to run `RAILS_ENV=test bundle exec rails db:create`

Is there an instruction set I'm not seeing for getting this to work?

evgeni commented:
@jcpunk are you by any chance using Fedora 43 (or any other distro that uses PostgreSQL 17 or newer), while running a PostgreSQL 13 container for Foreman?

evgeni commented:
I generated the snapshots locally, here is the diff:
```
diff --git test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host4and6dhcp.snap.txt test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host4and6dhcp.snap.txt
index 7fa3b6851..2bfafe2d6 100644
--- test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host4and6dhcp.snap.txt
+++ test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host4and6dhcp.snap.txt
@@ -20,8 +20,6 @@ firewall --service=ssh
 %end
 authconfig --useshadow --passalgo=sha512 --kickstart
 timezone --utc UTC --ntpservers first.ntp.server
-timesource --ntp-pool first.ntp-pool
-timesource --ntp-pool second.ntp-pool
 
 services --disabled gpm,sendmail,cups,pcmcia,isdn,rawdevices,hpoj,bluetooth,openibd,avahi-daemon,avahi-dnsconfd,hidd,hplip,pcscd
 
diff --git test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host4dhcp.snap.txt test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host4dhcp.snap.txt
index 01fe061b5..acf95df6c 100644
--- test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host4dhcp.snap.txt
+++ test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host4dhcp.snap.txt
@@ -20,8 +20,6 @@ firewall --service=ssh
 %end
 authconfig --useshadow --passalgo=sha512 --kickstart
 timezone --utc UTC --ntpservers first.ntp.server
-timesource --ntp-pool first.ntp-pool
-timesource --ntp-pool second.ntp-pool
 
 services --disabled gpm,sendmail,cups,pcmcia,isdn,rawdevices,hpoj,bluetooth,openibd,avahi-daemon,avahi-dnsconfd,hidd,hplip,pcscd
 
diff --git test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host4static.snap.txt test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host4static.snap.txt
index 86ab4c406..b769248fa 100644
--- test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host4static.snap.txt
+++ test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host4static.snap.txt
@@ -20,8 +20,6 @@ firewall --service=ssh
 %end
 authconfig --useshadow --passalgo=sha512 --kickstart
 timezone --utc UTC --ntpservers first.ntp.server
-timesource --ntp-pool first.ntp-pool
-timesource --ntp-pool second.ntp-pool
 
 services --disabled gpm,sendmail,cups,pcmcia,isdn,rawdevices,hpoj,bluetooth,openibd,avahi-daemon,avahi-dnsconfd,hidd,hplip,pcscd
 
diff --git test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host6dhcp.snap.txt test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host6dhcp.snap.txt
index f005f8817..f618b72eb 100644
--- test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host6dhcp.snap.txt
+++ test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host6dhcp.snap.txt
@@ -20,8 +20,6 @@ firewall --service=ssh
 %end
 authconfig --useshadow --passalgo=sha512 --kickstart
 timezone --utc UTC --ntpservers first.ntp.server
-timesource --ntp-pool first.ntp-pool
-timesource --ntp-pool second.ntp-pool
 
 services --disabled gpm,sendmail,cups,pcmcia,isdn,rawdevices,hpoj,bluetooth,openibd,avahi-daemon,avahi-dnsconfd,hidd,hplip,pcscd
 
diff --git test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host6static.snap.txt test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host6static.snap.txt
index 9a199d266..30ba75e5f 100644
--- test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host6static.snap.txt
+++ test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host6static.snap.txt
@@ -20,8 +20,6 @@ firewall --service=ssh
 %end
 authconfig --useshadow --passalgo=sha512 --kickstart
 timezone --utc UTC --ntpservers first.ntp.server
-timesource --ntp-pool first.ntp-pool
-timesource --ntp-pool second.ntp-pool
 
 services --disabled gpm,sendmail,cups,pcmcia,isdn,rawdevices,hpoj,bluetooth,openibd,avahi-daemon,avahi-dnsconfd,hidd,hplip,pcscd
 
diff --git test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.rhel9_dhcp.snap.txt test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.rhel9_dhcp.snap.txt
index af84c66a4..9b49628a3 100644
--- test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.rhel9_dhcp.snap.txt
+++ test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.rhel9_dhcp.snap.txt
@@ -18,7 +18,6 @@ firewall --service=ssh
 %addon com_redhat_kdump --disable
 %end
 authselect --useshadow --passalgo=sha512 --kickstart
-timezone --utc UTC
 timesource --ntp-pool first.ntp-pool
 timesource --ntp-pool second.ntp-pool
 
diff --git test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.rocky8_dhcp.snap.txt test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.rocky8_dhcp.snap.txt
index 772022d6d..5c853068c 100644
--- test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.rocky8_dhcp.snap.txt
+++ test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.rocky8_dhcp.snap.txt
@@ -19,8 +19,6 @@ firewall --service=ssh
 %end
 authselect --useshadow --passalgo=sha512 --kickstart
 timezone --utc UTC --ntpservers first.ntp.server
-timesource --ntp-pool first.ntp-pool
-timesource --ntp-pool second.ntp-pool
 
 services --disabled gpm,sendmail,cups,pcmcia,isdn,rawdevices,hpoj,bluetooth,openibd,avahi-daemon,avahi-dnsconfd,hidd,hplip,pcscd
 
diff --git test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.rocky9_dhcp.snap.txt test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.rocky9_dhcp.snap.txt
index db1e4701c..dcd8c0ec9 100644
--- test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.rocky9_dhcp.snap.txt
+++ test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.rocky9_dhcp.snap.txt
@@ -18,7 +18,6 @@ firewall --service=ssh
 %addon com_redhat_kdump --disable
 %end
 authselect --useshadow --passalgo=sha512 --kickstart
-timezone --utc UTC
 timesource --ntp-pool first.ntp-pool
 timesource --ntp-pool second.ntp-pool
 
```

is the `timezone` removal in EL9 correct?

jcpunk commented:
> @jcpunk are you by any chance using Fedora 43 (or any other distro that uses PostgreSQL 17 or newer), while running a PostgreSQL 13 container for Foreman?

For the snapshots I normally fire up a container with Almalinux 9 and follow this snippet: https://gist.github.com/jcpunk/e6bcd331ac95d0b7a97708abe5c2ddbb

jcpunk commented:
I think I may have sorted this.  As a note, https://pykickstart.readthedocs.io/en/latest/kickstart-docs.html#timezone show the timezone as the last argument. It probably doesn't matter but I updated the templates to match the documented entry exactly.

evgeni commented:
For the updated templates this seem to work fine, thank you!

I've updated the snapshots to use two servers using
```patch
diff --git app/services/foreman/template_snapshot_service.rb app/services/foreman/template_snapshot_service.rb
index 83a7bc68d..ed3fe6641 100644
--- app/services/foreman/template_snapshot_service.rb
+++ app/services/foreman/template_snapshot_service.rb
@@ -98,7 +98,7 @@ module Foreman
         "remote_desktop" => "true",
         "realm" => "true",
         "ntp-pools" => ['first.ntp-pool', 'second.ntp-pool'],
-        "ntp-server" => "first.ntp.server",
+        "ntp-server" => ["first.ntp.server", "second.ntp.server"],
       }
       host_params.each_pair do |name, value|
         FactoryBot.build(:host_parameter, host: host, name: name, value: value)
```

and see the following templates "broken" now:
```
diff --git test/unit/foreman/renderer/snapshots/ProvisioningTemplate/cloud-init/CloudInit_default.debian4dhcp.snap.txt test/unit/foreman/renderer/snapshots/ProvisioningTemplate/cloud-init/CloudInit_default.debian4dhcp.snap.txt
index c4e100076..d9590eb01 100644
--- test/unit/foreman/renderer/snapshots/ProvisioningTemplate/cloud-init/CloudInit_default.debian4dhcp.snap.txt
+++ test/unit/foreman/renderer/snapshots/ProvisioningTemplate/cloud-init/CloudInit_default.debian4dhcp.snap.txt
@@ -28,7 +28,7 @@ runcmd:
   
   echo "Updating system time"
               apt-get install -y chrony
-        sed -i -E 's|^pool .*|server first.ntp.server iburst|' /etc/chrony/chrony.conf
+        sed -i -E 's|^pool .*|server ["first.ntp.server", "second.ntp.server"] iburst|' /etc/chrony/chrony.conf
         systemctl enable --now chrony
         [ -f /usr/sbin/hwclock ] && /usr/sbin/hwclock --systohc
 - |
diff --git test/unit/foreman/renderer/snapshots/ProvisioningTemplate/cloud-init/CloudInit_default.ubuntu4dhcp.snap.txt test/unit/foreman/renderer/snapshots/ProvisioningTemplate/cloud-init/CloudInit_default.ubuntu4dhcp.snap.txt
index 81e07197d..3e260a118 100644
--- test/unit/foreman/renderer/snapshots/ProvisioningTemplate/cloud-init/CloudInit_default.ubuntu4dhcp.snap.txt
+++ test/unit/foreman/renderer/snapshots/ProvisioningTemplate/cloud-init/CloudInit_default.ubuntu4dhcp.snap.txt
@@ -27,7 +27,7 @@ runcmd:
 - |
   
   echo "Updating system time"
-              sed -i -E 's/^[[:space:]]*#?\s*NTP=.*/NTP=first.ntp.server/' /etc/systemd/timesyncd.conf
+              sed -i -E 's/^[[:space:]]*#?\s*NTP=.*/NTP=["first.ntp.server", "second.ntp.server"]/' /etc/systemd/timesyncd.conf
         systemctl enable --now systemd-timesyncd
         timedatectl set-ntp true
         [ -f /usr/sbin/hwclock ] && /usr/sbin/hwclock --systohc
diff --git test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Preseed_default_finish.debian4dhcp.snap.txt test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Preseed_default_finish.debian4dhcp.snap.txt
index 1ef97e80b..e4149a72e 100644
--- test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Preseed_default_finish.debian4dhcp.snap.txt
+++ test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Preseed_default_finish.debian4dhcp.snap.txt
@@ -4,7 +4,7 @@
 
 echo "Updating system time"
             apt-get install -y chrony
-      sed -i -E 's|^pool .*|server first.ntp.server iburst|' /etc/chrony/chrony.conf
+      sed -i -E 's|^pool .*|server ["first.ntp.server", "second.ntp.server"] iburst|' /etc/chrony/chrony.conf
       systemctl enable --now chrony
       [ -f /usr/sbin/hwclock ] && /usr/sbin/hwclock --systohc
 
diff --git test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Preseed_default_finish.ubuntu4dhcp.snap.txt test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Preseed_default_finish.ubuntu4dhcp.snap.txt
index 23993ad1f..f7532b87a 100644
--- test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Preseed_default_finish.ubuntu4dhcp.snap.txt
+++ test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Preseed_default_finish.ubuntu4dhcp.snap.txt
@@ -3,7 +3,7 @@
 
 
 echo "Updating system time"
-            sed -i -E 's/^[[:space:]]*#?\s*NTP=.*/NTP=first.ntp.server/' /etc/systemd/timesyncd.conf
+            sed -i -E 's/^[[:space:]]*#?\s*NTP=.*/NTP=["first.ntp.server", "second.ntp.server"]/' /etc/systemd/timesyncd.conf
       systemctl enable --now systemd-timesyncd
       timedatectl set-ntp true
       [ -f /usr/sbin/hwclock ] && /usr/sbin/hwclock --systohc
diff --git test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/AutoYaST_SLES_default.host4dhcp.snap.txt test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/AutoYaST_SLES_default.host4dhcp.snap.txt
index 877860722..454c8179b 100644
--- test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/AutoYaST_SLES_default.host4dhcp.snap.txt
+++ test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/AutoYaST_SLES_default.host4dhcp.snap.txt
@@ -28,7 +28,7 @@
     <configure_dhcp config:type="boolean">false</configure_dhcp>
     <peers config:type="list">
       <peer>
-        <address>first.ntp.server</address>
+        <address>["first.ntp.server", "second.ntp.server"]</address>
         <initial_sync config:type="boolean">true</initial_sync>
         <options></options>
         <type>server</type>
diff --git test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/AutoYaST_default.host4dhcp.snap.txt test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/AutoYaST_default.host4dhcp.snap.txt
index 06ed39091..a7261ee49 100644
--- test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/AutoYaST_default.host4dhcp.snap.txt
+++ test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/AutoYaST_default.host4dhcp.snap.txt
@@ -28,7 +28,7 @@
     <configure_dhcp config:type="boolean">false</configure_dhcp>
     <peers config:type="list">
       <peer>
-        <address>first.ntp.server</address>
+        <address>["first.ntp.server", "second.ntp.server"]</address>
         <initial_sync config:type="boolean">true</initial_sync>
         <options></options>
         <type>server</type>
```

using `.first` like you did for debian is probably the easiest fix for those?

jcpunk commented:
I think I got them all...

evgeni commented:
I don't think so, Tim. (Uh, what's better, Pokemon or Tool Time? Tough)

```diff
index 101c6ed9d..960abf2ed 100644
--- test/unit/foreman/renderer/snapshots/ProvisioningTemplate/cloud-init/CloudInit_default.debian4dhcp.snap.txt
+++ test/unit/foreman/renderer/snapshots/ProvisioningTemplate/cloud-init/CloudInit_default.debian4dhcp.snap.txt
@@ -28,7 +28,7 @@ runcmd:
   
   echo "Updating system time"
               apt-get install -y chrony
-                sed -i -E 's|^pool .*|server first.ntp.server iburst|' /etc/chrony/chrony.conf
+                sed -i -E 's|^pool .*|server first.ntp.server second.server iburst|' /etc/chrony/chrony.conf
               systemctl enable --now chrony
         [ -f /usr/sbin/hwclock ] && /usr/sbin/hwclock --systohc
 - |
diff --git test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Junos_default_finish.host4dhcp.snap.txt test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Junos_default_finish.host4dhcp.snap.txt
index 87e2968ce..b69be5dc2 100644
--- test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Junos_default_finish.host4dhcp.snap.txt
+++ test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Junos_default_finish.host4dhcp.snap.txt
@@ -10,7 +10,8 @@ system {
             }
     ntp {
         boot-server first.ntp.server;
-        server first.ntp.server;
+        boot-server second.server;
+        server ["first.ntp.server", "second.server"];
     }
     login {
         message "This device was provisioned by using The Foreman!\nSee http://theforeman.org/ for further information.\n";
diff --git test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Preseed_default_finish.debian4dhcp.snap.txt test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Preseed_default_finish.debian4dhcp.snap.txt
index ed5bc1892..884142c32 100644
--- test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Preseed_default_finish.debian4dhcp.snap.txt
+++ test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Preseed_default_finish.debian4dhcp.snap.txt
@@ -4,7 +4,7 @@
 
 echo "Updating system time"
             apt-get install -y chrony
-              sed -i -E 's|^pool .*|server first.ntp.server iburst|' /etc/chrony/chrony.conf
+              sed -i -E 's|^pool .*|server first.ntp.server second.server iburst|' /etc/chrony/chrony.conf
             systemctl enable --now chrony
       [ -f /usr/sbin/hwclock ] && /usr/sbin/hwclock --systohc
 
diff --git test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/XenServer_default_answerfile.host4dhcp.snap.txt test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/XenServer_default_answerfile.host4dhcp.snap.txt
index 1834a57db..5470fa8bd 100644
--- test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/XenServer_default_answerfile.host4dhcp.snap.txt
+++ test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/XenServer_default_answerfile.host4dhcp.snap.txt
@@ -12,7 +12,7 @@ part swap  --recommended  <keymap>us</keymap>
   </admin-interface>
   <name-server>192.168.42.2</name-server>
   <name-server>192.168.42.3</name-server>
-  <ntp-server>first.ntp.server</ntp-server>
+  <ntp-server>first.ntp.server second.server</ntp-server>
   <timezone>UTC</timezone>
   <time-config-method>ntp</time-config-method>
   <script stage="installation-complete" type="url">
```

[Xen seems to want](https://docs.xenserver.com/en-us/xenserver/8/install/advanced-install.html#ntp)
```
<ntp-server>one</ntp-server>
<ntp-server>two</ntp-server>
```
but I have no personal expirience here.

[chrony needs](https://chrony-project.org/doc/3.4/chrony.conf.html)
```
server one iburst
server two iburst
```

And [JunOS is confusing](https://www.juniper.net/documentation/us/en/software/junos/time-mgmt/topics/concept/synchronize-time-distribution-using-ntp.html), but probably doesn't want an stringified array.

Thanks for working on this!

jcpunk commented:
> I don't think so, Tim. (Uh, what's better, Pokemon or Tool Time? Tough)

blast...

> And [JunOS is confusing](https://www.juniper.net/documentation/us/en/software/junos/time-mgmt/topics/concept/synchronize-time-distribution-using-ntp.html), but probably doesn't want an stringified array.

I'm not seeing a rendering for JunOS.  Is `RAILS_ENV=test bundle exec rails snapshots:generate` the right thing to be using?

I'm going to abandon some of my more fancy loops in the cloud init space and just use `.first`.  I don't really have any great way to test that "on the ground".  I wanted to "do it right from the start" but I'm not sure I can make enough correct assumptions about file state and content. The `sed` I started working on got ugly...

evgeni commented:
Yeah, I updated ntp-server to an array in snapshot service, and regenerated the snapshots, which showed a change in a file with "junos" in the name.

And yes, .first sounds like the sensible fallback option here

jcpunk commented:
With luck, this time I've got it.

evgeni commented:
Yepp, looks good to me now!

evgeni commented:
@ekohl @ShimShtein anything on your side before we hit merge on this one?

jay7x commented on app/views/unattended/provisioning_templates/finish/freebsd_(mfsbsd)_finish.erb:
```diff
@@ -33,7 +33,11 @@ echo 'PermitRootLogin yes' >> /etc/ssh/sshd_config
 /bin/hostname <%= @host.name %>
 cp /usr/share/zoneinfo/<%= host_param('time-zone') || 'UTC' %> /etc/localtime
 adjkerntz -a
+<% if host_param('ntp-server').respond_to?(:join) -%>
+ntpdate <%= host_param('ntp-server').first %>
+<% else -%>
 ntpdate <%= host_param('ntp-server') || '0.freebsd.pool.ntp.org' %>
+<% end -%>
```
Can it be just like this?
```ruby
Array(host_param('ntp-server')).first || '0.freebsd.pool.ntp.org'
```

jcpunk replied:
I thought we tried that and safemode complained...

evgeni replied:
Yeah, pretty sure safemode will complain about that

ShimShtein replied:
We already discussed it https://github.com/theforeman/foreman/pull/10427#pullrequestreview-2910174625. Unfortunately we don't have it in safemode.

shubhamsg199 reviewed (approved):
Ack

ekohl commented on app/views/unattended/provisioning_templates/provision/kickstart_default.erb:
```diff
@@ -153,15 +153,29 @@ authconfig --useshadow --passalgo=<%= @host.operatingsystem.password_hash.downca
 <% elsif host_param('ntp-server') -%>
 timesource --ntp-server <%= host_param('ntp-server') %>
+<% else -%>
+timesource --ntp-server 0.pool.ntp.org
```
Do we need to hardcode a default timesource? I lean to leaving that out and rely on the built in mechanism as it was before.
```suggestion
```

jcpunk replied:
I put this in for consistency since so many of the other templates have a fall back to `0.pool.ntp.org`.

ekohl replied:
It's a tricky item. With others I wouldn't mind falling back to built in OS defaults rather than hardcoding, but that's beyond the scope of this PR IMHO.

With the eye on not changing too many things, I'd lean to not changing the current behavior.

@evgeni @ShimShtein @shubhamsg199 any thoughts?

ShimShtein replied:
I'm for merging as is. It is already tested and it's waiting for too long anyway. We can always fix it later, if it causes too many problems.

ekohl replied:
> I'm for merging as is. It is already tested and it's waiting for too long anyway. We can always fix it later, if it causes too many problems.

If anything, I'd go with matching the default in chrony and use the pool. https://chrony-project.org/examples.html has as the first line:
```
pool pool.ntp.org iburst
```

Reading https://pykickstart.readthedocs.io/en/latest/kickstart-docs.html#timesource that maps to using `--ntp-pool`:

```suggestion
<% else -%>
timesource --ntp-pool pool.ntp.org
```

There is actually a difference. https://www.ntppool.org/en/use.html has a sentence that's hard to parse:

>  Please also note that the system currently only provides IPv6 addresses for a zone in addition to IPv4 addresses if the zone name is prefixed by the number 2, e.g. 2.pool.ntp.org (provided there are any IPv6 NTP servers in the respective zone). Zone names not prefixed by a number, or prefixed with any of 0, 1 or 3, currently provide IPv4 addresses only.

A quick check tells me that `0.pool.ntp.org` indeed doesn't provide an IPv6 address while `2.pool.ntp.org` does.

jcpunk replied:
I've updated the patch series to use `2` for dual stack by default. It will probably need a squash, but I figured "showing my work" was probably beneficial here.

ekohl commented on app/views/unattended/provisioning_templates/provision/preseed_default.erb:
```diff
@@ -97,7 +97,16 @@ d-i time/zone string <%= host_param('time-zone') || 'UTC' %>
+<% end -%>
+<% else -%>
+d-i clock-setup/ntp-server string 0.debian.pool.ntp.org
+<% end -%>
```
This looks complex to me. Isn't this more consistent with the others and easier to read?
```suggestion
<% if host_param('ntp-server').respond_to?(:join) -%>
# host_param('ntp-server') is an array, but preseed only permits one value
d-i clock-setup/ntp-server string <%= host_param('ntp-server').first %>
<% else -%>
d-i clock-setup/ntp-server string <%= host_param('ntp-server') || '0.debian.pool.ntp.org' %>
<% end -%>
```

ekohl commented on app/views/unattended/provisioning_templates/provision/xenserver_default_answerfile.erb:
```diff
@@ -31,7 +31,15 @@ description: |
+<% elsif host_param('ntp-server') -%>
+  <ntp-server><%= host_param('ntp-server') %></ntp-server>
+<% else -%>
+  <ntp-server>pool.ntp.org</ntp-server>
```
This may be a personal preference, but I lean to keeping the template code to a minimum
```suggestion
<% else -%>
  <ntp-server><%= host_param('ntp-server') || 'pool.ntp.org' %></ntp-server>
```

ekohl commented on app/views/unattended/provisioning_templates/snippet/ntp.erb:
```diff
@@ -21,12 +21,20 @@ echo "Updating system time"
   <% if ntp_server -%>
     <% if @host.operatingsystem.name == "Ubuntu" -%>
-      sed -i -E 's/^[[:space:]]*#?\s*NTP=.*/NTP=<%= ntp_server %>/' /etc/systemd/timesyncd.conf
+      <% if host_param('ntp-server').respond_to?(:join) -%>
```
```suggestion
      <% if ntp_server.respond_to?(:join) -%>
```

ekohl commented on app/views/unattended/provisioning_templates/snippet/ntp.erb:
```diff
@@ -21,12 +21,20 @@ echo "Updating system time"
     <% else -%>
       apt-get install -y chrony
-      sed -i -E 's|^pool .*|server <%= ntp_server %> iburst|' /etc/chrony/chrony.conf
+      <% if host_param('ntp-server').respond_to?(:join) -%>
```
```suggestion
      <% if ntp_server.respond_to?(:join) -%>
```

ekohl commented on app/views/unattended/provisioning_templates/snippet/ntp.erb:
```diff
@@ -39,7 +47,11 @@ echo "Updating system time"
     yum -y install ntpdate
     <% if host_param('ntp-server') -%>
-      /usr/sbin/ntpdate -sub <%= host_param('ntp-server') %>
+      <% if host_param('ntp-server').respond_to?(:join) -%>
```
```suggestion
      <% if ntp_server.respond_to?(:join) -%>
```

ekohl reviewed (commented):
I must admit I didn't review the whole conversation. Inline I made a few comments. Nothing serious, but more focused on keeping it readable.

ekohl commented on app/views/unattended/provisioning_templates/provision/kickstart_default.erb:
```diff
@@ -153,15 +153,29 @@ authconfig --useshadow --passalgo=<%= @host.operatingsystem.password_hash.downca
-timezone --utc <%= host_param('time-zone') || 'UTC' %> <%= host_param('ntp-server') ? "--ntpservers #{host_param('ntp-server')}" : '' %>
+<%# RHEL < 9: Use legacy timezone command with --ntpservers -%>
+<% if host_param('ntp-server').respond_to?(:join) -%>
+timezone --ntpservers <%= host_param('ntp-server').join(',') %> --utc <%= host_param('time-zone') || 'UTC' %>
+<% elsif host_param('ntp-server') -%>
+timezone --ntpservers <%= host_param('ntp-server') %> --utc <%= host_param('time-zone') || 'UTC' %>
```
Slight nit: if you don't change the order of `--ntpservers` and `--utc` then you don't need as much snapshot changes. Probably doesn't really matter but smaller changes are always easier.

jcpunk replied:
The anaconda documentation lists them in the order I've got here. It probably doesn't matter but....

ShimShtein commented on test/unit/foreman/renderer/snapshots/ProvisioningTemplate/cloud-init/CloudInit_default.debian4dhcp.snap.txt:
```diff
@@ -28,8 +28,8 @@ runcmd:
               apt-get install -y chrony
-        sed -i -E 's|^pool .*|server first.ntp.server iburst|' /etc/chrony/chrony.conf
-        systemctl enable --now chrony
+                sed -i -E 's|^pool .*|server first.ntp.server iburst|' /etc/chrony/chrony.conf
```
cloud-init templates are quite sensitive to extra indentation (specifically for the first line in multiline values). While this is not the first line, I would ask to remove the extra spaces, even if it means uglier template.

ShimShtein commented on test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/AutoYaST_default.host4dhcp.snap.txt:
```diff
@@ -28,8 +28,8 @@
       <peer>
-        <address>first.ntp.server</address>
-        <initial_sync config:type="boolean">true</initial_sync>
+                <address>first.ntp.server</address>
```
This is totally nitpicking, but if you already work on spacing, could you please remove the extra spacing in the relevant template?
This is less critical, since the file is in XML format.

jcpunk commented:
I think I've got the templates better aligned.  Does this look workable?

ShimShtein commented:
@ekohl I'm OK with merging this PR. Using dualstack as a default seems OK for me. Any comments?

ekohl reviewed (approved):
I think this addresses my concerns.
